apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js

```diff
@@ -35,6 +34,34 @@ const editSitePackage = require( '@wordpress/edit-site' );
 
 const debug = debugFactory( 'wpcom-block-editor:iframe-bridge-server' );
 
+const clickOverrides = {};
+let addedListener = false;
+// Replicates basic '$( el ).on( selector, cb )' with preventDefault
```

Comment:
This does not really have the same behaviour as the code it's replacing, since it won't pick up new elements that get added to the document after the method runs.

Whether or not that can happen here is unclear to me, but since we're dealing with `react`, which takes over much of the DOM and does who knows what to the elements inside it, I'd err on the side of caution.

To replicate `jQuery`'s behaviour here, you'll need to instead add an event listener to something like `document` or `body`, handle the event during the capture phase, have it match the target element against the selector, and apply it to the correct element.

I've usually found it easier to refactor the existing code and come up with a specific solution for each case, rather than try to design a generic replacement like this.

Reply:
> you'll need to instead add an event listener to something like...

This is exactly what I attempted to accomplish, but I'd love to hear more if not :) The event listener should run on clicks inside `#editor`, and then dynamically check if the event happens inside an element matching the selector. It should match elements at the time of the event, not at the time of creation. (We could use a different root element easiy -- `#editor` is what jquery used, and it does seem to work here.)

This function is oddly named, because I originally didn't use the approach you mention. This function attaches a root listener if it doesn't exist and also stores the selector and callback for reference on click.

While I do like the idea of coming up with more specific solutions, each case here is basically the same thing: on click to a link/button in the editor, do a Calypso-specific thing _instead_ of a wp-admin specific thing. (So, open Calypso revisions in a modal instead of redirecting to `revisions.php`, or redirect to the calypso customizer instead of the wp-admin customizer)

And often, there is more than one button/link to match. So, for every link to `revisions.php`, it needs to do this other thing instead :)

I don't think there are alternative ways to extend these buttons or links, but I'm open to ideas :) One idea is to rewrite the way Gutenberg does redirects. So when it wants to redirect to a URL, it checks to see if an extension wants to change that URL or do something else. But I'm not even sure that would be a welcome contribution to Gutenberg.

Reply:
> This is exactly what I attempted to accomplish, but I'd love to hear more if not :)

Ugh, you're totally right. I missed that when reviewing this earlier. Sorry about that! 😞

> It should match elements at the time of the event, not at the time of creation.

Yes, it definitely should do that 👍 However, and I could be wrong, I don't think that `preventDefault` is guaranteed to work correctly in this situation. Since this event handler is being registered for the bubbling phase (no `useCapture` or `options.capture` parameter), and since it's being registered at the container element level, the default actions on the contained elements will be triggered before bubbling up to the next level, as I understand.

As such, you'd need to register the event listener on the capture phase instead, while the DOM is being traversed in the opposite direction.

Again, I could be wrong, and I haven't tested things in this PR, but I believe that's how it's supposed to work.

> This function is oddly named, because I originally didn't use the approach you mention. This function attaches a root listener if it doesn't exist and also stores the selector and callback for reference on click.

How about something like `addEditorListener` or `addProxiedListener` instead? `toEach` strongly implies the idea of a loop to me.

> One idea is to rewrite the way Gutenberg does redirects. So when it wants to redirect to a URL, it checks to see if an extension wants to change that URL or do something else. But I'm not even sure that would be a welcome contribution to Gutenberg.

Given the constraints, I agree that this is probably the best solution 👍

Reply:
Good question about prevent default. It tests okay, in that it opens the Calypso revisions modal without redirecting to revisions.php. But I admit that I don't know much about how event bubbling works. 

Also, great name ideas.

> I agree that this is probably the best solution

By "this", do you mean the gutenberg rework or this PR? 😄

Reply:
> By "this", do you mean the gutenberg rework or this PR? 😄

Haha, good point! I meant this PR 🙂

Reply:
> Good question about prevent default. It tests okay, in that it opens the Calypso revisions modal without redirecting to revisions.php. But I admit that I don't know much about how event bubbling works.

If it tests well cross-browser, then it's probably fine and you can ignore my comment 😄